lib/compat/wordpress-6.0/register-webfonts-from-theme-json.php

```diff
@@ -12,15 +12,35 @@ function gutenberg_register_webfonts_from_theme_json() {
 	// Get settings from theme.json.
 	$theme_settings = WP_Theme_JSON_Resolver_Gutenberg::get_theme_data()->get_settings();
 
+	// Get variations.
+	$variations = gutenberg_get_theme_styles_variations();
+
+	// Merge settings with variations.
+	$settings = $theme_settings;
+	foreach ( $variations as $variation ) {
+
+		// Sanity check: Skip if fontFamilies are not defined in the variation.
+		if (
+			empty( $variation['settings'] ) ||
+			empty( $variation['settings']['typography'] ) ||
+			empty( $variation['settings']['typography']['fontFamilies'] )
+		) {
+			continue;
+		}
+		$settings['typography']                 = empty( $settings['typography'] ) ? array() : $settings['typography'];
+		$settings['typography']['fontFamilies'] = empty( $settings['typography']['fontFamilies'] ) ? array() : $settings['typography']['fontFamilies'];
+		$settings['typography']['fontFamilies'] = array_merge( $settings['typography']['fontFamilies'], $variation['settings']['typography']['fontFamilies'] );
+	}
```

Comment:
If I'm following this properly, this basically just merges the font families from the variations into the main theme.json output. 

While this may work for now, this might not be the best approach I think. Our goal is to be able to load variations from potentially the directory (theme or dedicated directory) and feel it would be better if instead of doing this, we provide a way for the site editor to lazy-load fonts for dedicated variations.

Reply:
I'm trying to think of a way to "lazy-load" these on the frontend, but I can't think of something... From what I can tell, we don't add a body-class or something that would help us identify whether a variation is used or not, correct?

Reply:
Yeah! there's no concept of "active variation", but maybe that's an advantage instead of being a problem. Basically 

We have this in the site editor:

```js
<GlobalStylesProvider config={ config }>
   <Iframe>
      <SiteEditor />
    </Iframe>
</GlobaStylesProvider />
```

and for each style variation on the style variation picker we do something like this as well:

```
<GlobalStylesProvider config={ config }>
  <IFrame> 
   <VariationPreview />
   </IFrame>
</GlobaStylesProvider />
```

So both places have the same structure, so we can use the global styles provider config to get the font family and call a URL to load these fonts in the iframe or something like that.

Would something like that work?

Reply:
tbh, I'm more concerned with the frontend and not the editor...We can probably find a way to handle it on the editor, but on the frontend we'll still need a way to figure out what the "active variation" is in order to load the appropriate variation JSON files and then parse the webfonts from that.
That's what troubles me right now... 
If variation JSON files simply pick things that are defined in the main JSON file like colors, font-sizes and whatnot, we don't need to know what the active variation is. But if in that file we're actually defining new things, then we'll need to know which one is active :hurtrealbad:

Reply:
> We can probably find a way to handle it on the editor, but on the frontend we'll still need a way to figure out what the "active variation" is in order to load the appropriate variation JSON files and then parse the webfonts from that.

No, in the frontend there's no problem because the variation will be saved as a "user theme.json" so I think there won't be any issue, it should already be working.

Reply:
Yep, that worked! 👍 I just needed to change `get_theme_data()` to `get_merged_data()`.

As for lazy-loading assets in the editor, I got the impression from https://github.com/WordPress/gutenberg/issues/39332#issuecomment-1085157058 that we don't want to do that at this stage, and it's something we'll want to do later-on?

Reply:
No I think that comment is talking about something else: Basically there's a difference between all registered fonts and all fonts that a theme use (theme.json or variations). For the former we don't want to enqueue them anywhere, for the latter, we want to enqueue them for both frontend and editor.

The variation is special in the sense that in the editor, we only want to load them in the iframe showing the preview of the variation unless the variation is applied as "user theme.json" in which case it should be loaded in the main iframe too.

Reply:
>  Basically there's a difference between all registered fonts and all fonts that a theme use (theme.json or variations). For the former we don't want to enqueue them anywhere, for the latter, we want to enqueue them for both frontend and editor.

It [sounds like ](https://github.com/WordPress/gutenberg/issues/39332#issuecomment-1085157058)we may be conflicting on this: "The main goal is to avoid loading every registered font on the front-end while still loading them in the site editor (back-end) so people can choose and manage them."

Reply:
@Addison-Stavlo that's why I converted this PR to a draft... I'm still a bit unclear on what it is we want to achieve.
The frontend now works as expected in this PR, but on the editor things got complicated and I don't know how to address it 😞 
If we want to **always** load the webfonts, it's relatively easy to do. If we want to "lazy-load" them, then I don't know how to properly address it.

Reply:
> If we want to always load the webfonts, it's relatively easy to do

Maybe we can start with that first but I imagine it's not sustainable for the future if we want to load all "registered" fonts (not particularly variations ones but all the ones that are in google fonts for instance).

Reply:
OK. For now I suppose it's a reasonable compromise. Visitors will only get the font if they're using it, while editors will get more fonts enqueued.

I think that we'll need a more generic mechanism to lazy-load stuff in the editor anyway, so this could probably be a part of a more generalized approach 👍